### PR TITLE
fix(RHINENG-11769) Fix Export Data Field names and add some new

### PR DESCRIPTION
**Pull request by strider**
# Overview

This PR is being created to address [RHINENG-11769](https://issues.redhat.com/browse/RHINENG-11769).

This commit adds a new host serializer specifically for the exp-svc feature.

Example:

```
{'display_name': 'b1280d.foo.redhat.com',
     'group_id': None,
     'group_name': None,
     'host_type': 'conventional',
     'host_id': '937c265c-53a3-40c5-9a83-cbc5bf913a27',
     'fqdn': 'b1280d.foo.redhat.com',
     'os_release': 'Red Hat EL 7.0.1',
     'satellite_id': None,
     'state': 'fresh',
     'subscription_manager_id': None,
     'updated': '2024-08-07T12:09:54.867575+00:00',
     'tags': [{'key': 'key3', 'namespace': 'NS1', 'value': 'val3'},
              {'key': 'key3', 'namespace': 'NS3', 'value': 'val3'},
              {'key': 'prod', 'namespace': 'Sat', 'value': None},
              {'key': 'key', 'namespace': 'SPECIAL', 'value': 'val'}]}
```

## PR Checklist

- [ ] Keep PR title short, ideally under 72 characters
- [ ] Descriptive comments provided in complex code blocks
- [ ] Include raw query examples in the PR description, if adding/modifying SQL query
- [ ] Tests: validate optimal/expected output
- [ ] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Uses [type hinting](https://docs.python.org/3/library/typing.html), if convenient
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

**Review by thearifismail — COMMENTED** [review]
I am not going to block this PR but IMO, the requirements doc should be fixed to state `os_release` and `modified_on`, which is what in the code and better represent the state of host objects than using `os_version` and `ast_updated`

File: app/serialization.py [hunk 1]
```diff
@@ -246,13 +254,19 @@ def serialize_host(
...
     if "state" in fields:
         serialized_host["state"] = Conditions.find_host_state(
             stale_timestamp=stale_timestamp, stale_warning_timestamp=stale_warning_timestamp
         )
     if "host_type" in fields:
         serialized_host["host_type"] = host.host_type
+        if for_export_svc and not host.host_type:
+            # For export service, host_type should be exported as conventional
+            # if the host is not an edge one instead of None.
+            serialized_host["host_type"] = "conventional"
```
[the remark this answers is not in the document]
Comment (strider) [reply app/serialization.py:269]: This is only a formatting purpose and we do this only for the export service. The schema should not be changed. We just want to avoid the `None` value in the export archive.

File: app/serialization.py [hunk 2]
```diff
@@ -180,10 +182,14 @@ def serialize_host(
 
     if "id" in fields:
         serialized_host["id"] = _serialize_uuid(host.id)
+        if for_export_svc:
+            serialized_host["host_id"] = serialized_host["id"]
     if "account" in fields:
         serialized_host["account"] = host.account
     if "org_id" in fields:
         serialized_host["org_id"] = host.org_id
+    if "fqdn" in fields and for_export_svc:
```
[the remark this answers is not in the document]
Comment (strider) [reply app/serialization.py:191]: I agree that a better management of this `fqdn` exported field should be done here. I am going to propose something else according to your recommandation.

PS: This is only for the export service and how we format the exported data for the customer.

File: app/serialization.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (thearifismail) [reply app/serialization.py:271]: Don't want to block this but is important in my opinion to be dealt with.  There was another instance where `host_type=None` made it awkward to deal with.  To address the issue, I just submitted [RHINENG-11889](https://issues.redhat.com/browse/RHINENG-11889)

**Review by thearifismail — DISMISSED** [review]

**Review by jpramos123 — DISMISSED** [review]

File: app/serialization.py [hunk 3]
```diff
@@ -247,13 +258,19 @@ def serialize_host(
             serialized_host["os_release"] = host.system_profile_facts["os_release"]
         else:
             serialized_host["os_release"] = None
+        if for_export_svc:
+            serialized_host["os_version"] = serialized_host["os_release"]
```
Comment (kruai) [inline app/serialization.py:262]: Doing it this way will result in the serialized host having both "os_version" and "os_release" fields. Same goes for the rest of these fields, so I won't comment on all of them. Is the idea to put all of them (including duplicate fields) inside the serialized_host object, and then remove all the ones that aren't needed for the export service at the end?

File: app/serialization.py [hunk 4]
```diff
@@ -211,6 +220,8 @@ def serialize_host(
         serialized_host["created"] = _serialize_datetime(host.created_on)
     if "updated" in fields:
         serialized_host["updated"] = _serialize_datetime(host.modified_on)
+        if for_export_svc:
+            serialized_host["last_updated"] = serialized_host["updated"]
```
Comment (kruai) [inline app/serialization.py:224]: Doing it this way will result in the serialized host having both "last_updated" and "updated" fields

**Review by kruai — COMMENTED** [review]
This technically works, but these changes are complicating the serialization logic even further, especially now that export-service uses different names than our usual serialization. I'd really prefer if we could separate out the export-service-specific logic

File: app/serialization.py [hunk 5]
```diff
@@ -22,14 +22,15 @@
...
+    "hostname",
     "subscription_manager_id",
     "satellite_id",
     "display_name",
     "group_id",
     "group_name",
-    "os_release",
-    "updated",
+    "os_version",
+    "last_updated",
```
Comment (thearifismail) [inline app/serialization.py:33]: Couldn't `udpated` be used instead of creating a new attribute, `last_updated`?  Furthermore, even `updated` should be replaced by `modified_on`, which is what is used as the DB column title.

File: app/serialization.py (hunk 3 above)
Comment (thearifismail) [reply app/serialization.py:262]: `os_version` should replace `os_release`.

**Review by thearifismail — COMMENTED** [review]
I have the same comment as @kruai.  `last_updated` should be replaced by `updated` or `modified_on`.  `os_release` is not needed.  Thanks to anybody can deliver to any project, the System Profile schema has quite a few overlapping fields which are responsible for this confusing situation.

**Comment by jpramos123** [comment]
@kruai @thearifismail I created a new host serializer specifically for the exp-svc feature

File: app/serialization.py (hunk 5 above)
Comment (jpramos123) [reply app/serialization.py:33]: We want to use the fields specified in this card: https://issues.redhat.com/browse/RHINENG-11769

File: app/serialization.py (hunk 3 above)
Comment (jpramos123) [reply app/serialization.py]: @thearifismail We want to use the fields specified in this card: https://issues.redhat.com/browse/RHINENG-11769

File: app/serialization.py (hunk 5 above)
Comment (thearifismail) [reply app/serialization.py:33]: IMO, the fields specified in [RHINENG-11769](https://issues.redhat.com/browse/RHINENG-11769) should be changed to "updated".  I am sure @msager will not have any problem unless "updated" is being used already for some other reason and is still required.  My intent is to not add another variable that we could do without.

Comment (thearifismail) [reply app/serialization.py:33]: IMO, the fields specified in [RHINENG-11769](https://issues.redhat.com/browse/RHINENG-11769) should be changed to "updated".  I am sure @msager will not have any problem unless "updated" is being used already for some other reason and is still required.  My intent is to not add another variable that we could do without.

Comment (thearifismail) [reply app/serialization.py:33]: IMO, the fields specified in [RHINENG-11769](https://issues.redhat.com/browse/RHINENG-11769) should be changed to "updated".  I am sure @msager will not have any problem unless "updated" is being used already for some other reason and is still required.  My intent is to not add another variable that we could do without.

Comment (thearifismail) [reply app/serialization.py:33]: IMO, the fields specified in [RHINENG-11769](https://issues.redhat.com/browse/RHINENG-11769) should be changed to "updated".  I am sure @msager will not have any problem unless "updated" is being used already for some other reason and is still required.  My intent is to not add another variable that we could do without.

Comment (thearifismail) [reply app/serialization.py:33]: IMO, the fields specified in [RHINENG-11769](https://issues.redhat.com/browse/RHINENG-11769) should be changed to "updated".  I am sure @msager will not have any problem unless "updated" is being used already for some other reason and is still required.  My intent is to not add another variable that we could do without.

File: app/serialization.py [hunk 6]
```diff
@@ -255,9 +247,77 @@ def serialize_host(
     if "host_type" in fields:
         serialized_host["host_type"] = host.host_type
 
-    if for_export_svc:
-        serialized_host = {key: serialized_host[key] for key in _EXPORT_SERVICE_FIELDS}
+    return serialized_host
+
+
+def serialize_host_export_svc(
```
Comment (thearifismail) [inline app/serialization.py:253]: the function name does read right; it is a host that is being serialized NOT a service.  How about `serialize_host_to_export()`?

File: app/serialization.py (hunk 3 above)
Comment (FabriciaDinizRH) [reply app/serialization.py]: @jpramos123 the problem that Asa mentioned is still in the code, about having the two fields present at the end.

Comment (jpramos123) [reply app/serialization.py]: @FabriciaDinizRH I don't see this anymore...
If we take a look at these lines: https://github.com/RedHatInsights/insights-host-inventory/pull/1869/files#diff-957087e16a2fa0e73c463f90ad900d1d451a760f3d0a4c4f1ac461e76ed6626cR304-R308

We are using the `os_release` that is in the `field` variable to create the `os_version` that is in the `serialized_host` dict. So there no duplication for this field

**Review by thearifismail — COMMENTED** [review]
please see my comments

File: app/serialization.py (hunk 6 above)
Comment (kruai) [reply app/serialization.py:253]: Or `serialize_host_for_export` / `serialize_host_for_export_svc`

File: app/serialization.py (hunk 5 above)
Comment (kruai) [reply app/serialization.py:33]: I agree with Arif that we should use the existing field names whenever possible, assuming we have a choice here

**Review by kruai — CHANGES_REQUESTED** [review]

**Comment by thearifismail** [comment]
I am sorry. I thought you had made the changes.
- Arif

File: app/serialization.py [hunk 7]
```diff
@@ -22,7 +22,8 @@
 
 
 _EXPORT_SERVICE_FIELDS = [
-    "id",
+    "host_id",
+    "hostname",
```
Comment (kruai) [inline app/serialization.py:26]: I opened some discussion on this [here](https://docs.google.com/document/d/1MyfSjPpbu-E0oYPPGRsV-S9Mor1FbMqcv-tlDGUCRFE/edit); I'd like if we can just keep using "fqdn" rather than "hostname"

Comment (jpramos123) [reply app/serialization.py]: Changed `hostnane` to use the default `fqdn` field

File: tests/helpers/export_service_utils.py [hunk 8]
```diff
@@ -56,7 +59,7 @@
             {"key": "prod", "namespace": "Sat", "value": None},
             {"key": "key", "namespace": "SPECIAL", "value": "val"},
         ],
-        "updated": "2024-07-23T16:24:41.167465+00:00",
+        "last_updated": "2024-07-23T16:24:41.167465+00:00",
```
Comment (thearifismail) [inline tests/helpers/export_service_utils.py:62]: this "last_updated" should be replaced by "updated".

File: tests/helpers/export_service_utils.py [hunk 9]
```diff
@@ -20,15 +21,16 @@
             {"key": "prod", "namespace": "Sat", "value": None},
             {"key": "key", "namespace": "SPECIAL", "value": "val"},
         ],
-        "updated": "2024-07-23T16:24:41.159435+00:00",
+        "last_updated": "2024-07-23T16:24:41.159435+00:00",
```
Comment (thearifismail) [inline tests/helpers/export_service_utils.py:24]: this "last_updated" should be replaced by "updated".

File: tests/helpers/export_service_utils.py [hunk 10]
```diff
@@ -38,15 +40,16 @@
             {"key": "prod", "namespace": "Sat", "value": None},
             {"key": "key", "namespace": "SPECIAL", "value": "val"},
         ],
-        "updated": "2024-07-23T16:24:40.562889+00:00",
+        "last_updated": "2024-07-23T16:24:40.562889+00:00",
```
Comment (thearifismail) [inline tests/helpers/export_service_utils.py:43]: this "last_updated" should be replaced by "updated".

File: tests/helpers/export_service_utils.py (hunk 8 above)
Comment (thearifismail) [inline tests/helpers/export_service_utils.py:62]: this "last_updated" should be replaced by "updated".

File: app/serialization.py [hunk 11]
```diff
@@ -235,29 +233,71 @@ def serialize_host(
...
+        stale_warning_timestamp = staleness_timestamps.stale_warning_timestamp(
+            host.modified_on, staleness["immutable_time_to_stale_warning"]
+        )
+    else:
+        stale_timestamp = staleness_timestamps.stale_timestamp(
+            host.modified_on, staleness["conventional_time_to_stale"]
+        )
+        stale_warning_timestamp = staleness_timestamps.stale_warning_timestamp(
+            host.modified_on, staleness["conventional_time_to_stale_warning"]
+        )
```
Comment (kruai) [inline app/serialization.py:270]: This logic is unnecessary here. `serialized_host` has those timestamps on it when it gets returned from `serialize_host()`, so there's no need to duplicate the code

File: app/serialization.py [hunk 12]
```diff
@@ -235,29 +233,73 @@ def serialize_host(
...
+
+
+def serialize_host_for_export_svc(
+    host,
+    staleness_timestamps,
+    staleness=None,
+):
+    # TODO: In future, this must handle groups staleness
+
+    if host.host_type == "edge" or (
```
[the remark this answers is not in the document]
Comment (thearifismail) [reply app/serialization.py:255]: Will this question be answered in this PR?

File: app/serialization.py [hunk 13]
```diff
@@ -235,29 +233,71 @@ def serialize_host(
...
+            host.modified_on, staleness["conventional_time_to_stale_warning"]
+        )
+
+    serialized_host = serialize_host(
+        host, staleness_timestamps=staleness_timestamps, staleness=staleness, additional_fields=("os_release",)
+    )
+
+    fields = DEFAULT_FIELDS + ADDITIONAL_EXPORT_SERVICE_FIELDS
+
+    if "id" in fields:
```
Comment (kruai) [inline app/serialization.py:278]: `fields` is constant, so there's no need to check for "id", "groups", etc. You can remove all of those checks

File: tests/helpers/export_service_utils.py [hunk 14]
```diff
@@ -6,11 +6,12 @@
...
         "display_name": "b4f637.foo.redhat.com",
+        "hostname": "b4f637.foo.redhat.com",
         "group_id": None,
         "group_name": None,
-        "host_type": None,
-        "id": "80e3f012-c8e9-4435-8d84-307ff85536e4",
-        "os_release": "Red Hat EL 7.0.1",
+        "host_type": "conventional",
+        "host_id": "80e3f012-c8e9-4435-8d84-307ff85536e4",
+        "os_version": "Red Hat EL 7.0.1",
```
Comment (kruai) [inline tests/helpers/export_service_utils.py:14]:
```suggestion
        "os_release": "Red Hat EL 7.0.1",
```

File: tests/helpers/export_service_utils.py (hunk 9 above)
Comment (kruai) [inline tests/helpers/export_service_utils.py:24]:
```suggestion
        "updated": "2024-07-23T16:24:41.159435+00:00",
```

File: tests/helpers/export_service_utils.py [hunk 15]
```diff
@@ -20,15 +21,16 @@
...
         "display_name": "063406.foo.redhat.com",
+        "hostname": "063406.foo.redhat.com",
         "group_id": None,
         "group_name": None,
-        "host_type": None,
-        "id": "c1eb8355-b00c-4b0e-af59-c691a0f88629",
-        "os_release": "Red Hat EL 7.0.1",
+        "host_type": "conventional",
+        "host_id": "c1eb8355-b00c-4b0e-af59-c691a0f88629",
+        "os_version": "Red Hat EL 7.0.1",
```
Comment (kruai) [inline tests/helpers/export_service_utils.py:33]:
```suggestion
        "os_release": "Red Hat EL 7.0.1",
```

File: tests/helpers/export_service_utils.py (hunk 10 above)
Comment (kruai) [inline tests/helpers/export_service_utils.py:43]:
```suggestion
        "updated": "2024-07-23T16:24:40.562889+00:00",
```

File: tests/helpers/export_service_utils.py [hunk 16]
```diff
@@ -38,15 +40,16 @@
...
         "display_name": "43794f.foo.redhat.com",
+        "hostname": "43794f.foo.redhat.com",
         "group_id": None,
         "group_name": None,
-        "host_type": None,
-        "id": "c0f828f6-5efe-476e-8abc-be4e93c4098c",
-        "os_release": "Red Hat EL 7.0.1",
+        "host_type": "conventional",
+        "host_id": "c0f828f6-5efe-476e-8abc-be4e93c4098c",
+        "os_version": "Red Hat EL 7.0.1",
```
Comment (kruai) [inline tests/helpers/export_service_utils.py:52]:
```suggestion
        "os_release": "Red Hat EL 7.0.1",
```

File: tests/helpers/export_service_utils.py (hunk 8 above)
Comment (kruai) [inline tests/helpers/export_service_utils.py:62]:
```suggestion
        "updated": "2024-07-23T16:24:41.167465+00:00",
```

**Review by kruai — CHANGES_REQUESTED** [review]

**Review by kruai — APPROVED** [review]

File: app/serialization.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jpramos123) [reply app/serialization.py]: Removed that portion of the code that is not necessary here. 

Thanks

**Review by thearifismail — APPROVED** [review]
